Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -0,0 +1,215 @@
+/// This cache is intended to be written to and read from using the same serial queue.
+public final class CacheAdvance<T: Codable> {
+
+    // MARK: Initialization
+
+    /// Creates a new instance of the receiver.
+    ///
+    /// - Parameters:
+    ///   - file: The file URL indicating the desired location of the on-disk store. This file should already exist.
+    ///   - maximumBytes: The maximum size of the cache, in bytes. Logs larger than this size will fail to append to the store.
+    ///   - shouldOverwriteOldMessages: When `true`, once the on-disk store exceeds maximumBytes, new entries will replace the oldest entry.
+    ///
+    /// - Warning: `maximumBytes` must be consistent for the life of a cache. Changing this value after logs have been persisted to a cache leads to undefined behavior.
+    public init(
+        file: URL,
+        maximumBytes: Bytes,
+        shouldOverwriteOldMessages: Bool)
+        throws
+    {
+        self.file = file
+        self.maximumBytes = maximumBytes
+        self.shouldOverwriteOldMessages = shouldOverwriteOldMessages
+
+        writer = try FileHandle(forWritingTo: file)
+        reader = try FileHandle(forReadingFrom: file)
+
+        lengthOfMessageSuffix = {
+            if shouldOverwriteOldMessages {
+                /// The message suffix requires space for both `endOfNewestMessageMarker` and `offsetOfFirstMessage` after each message when rolling is enabled.
+                return Bytes(Data.endOfNewestMessageMarker.count) + Bytes(Data.bytesStorageLength)
+            } else {
+                /// The message suffix requires space for `endOfNewestMessageMarker` after each message.
+                return Bytes(Data.endOfNewestMessageMarker.count)
+            }
+        }()
+    }
+
+    deinit {
+        try? writer.close()
+        try? reader.close()
+    }
+
+    // MARK: Public
+
+    public let file: URL
+
+    /// Appends a message to the cache.
+    /// - Parameter message: A message to write to disk.
+    public func append(message: T) throws {
+        try setUpFileHandlesIfNecessary()
+
+        let encodedMessage = EncodableMessage(message: message, encoder: encoder)
+        let messageData = try encodedMessage.encodedData()
+        let messageLength = Bytes(messageData.count)
+
+        guard messageLength > 0 else {
+            // The message length has the same value as as our `endOfNewestMessageMarker`.
+            // Storing this message could cause data corruption by fooling the cache into thinking the prior message is the last in the cache.
+            // JSON values (even empty ones) are always encoded with length > 0. If this condition is hit, the message is clearly corrupt.
+            throw CacheAdvanceWriteError.messageCorrupted
+        }
+
+        let bytesNeededToStoreMessage = messageLength + lengthOfMessageSuffix
+        guard bytesNeededToStoreMessage <= maximumBytes && bytesNeededToStoreMessage < Bytes(MessageSpan.max) else {
```

Comment:
> It seems like we'd never want to be in a situation where `MessageSpan.max` is greater than `maximumBytes`.

I think you mean the inverse? `MessageSpan.max` is `UInt32.max` (i.e. four gigabytes), and `maximumBytes` is the maximum byte size of this cache, which is likely to be pretty small by comparison.

`MessageSpan` is an `UInt32` and `Bytes` is `UInt64`. I wanted to enable _massive_ maximum caches for those that wanted it, which means that it is entirely possible for `maximumBytes` to greatly exceed a `MessageSpan`'s capability. It is currently user error to log a single message larger than four gigabytes. Using a `UInt32` to define the `MessageSpan` allows me to store the length of each message in half the space vs the space if I'd used an `UInt64`. Martin suggested earlier in this PR that I make `MessageSpan` a generic so that folk can write messages longer than `UInt32.max` in length if they want to opt-into doing so. I think we can tackle that kind of customizability once we have a header or file metadata stored somewhere.